velox/expression/fuzzer/ExpressionFuzzer.cpp

```diff
@@ -1004,23 +1004,41 @@ core::TypedExprPtr ExpressionFuzzer::generateExpression(
       chosenFunctionName = templateList[chosenExprIndex];
     }
 
-    if (chosenFunctionName == "cast") {
-      expression = generateCastExpression(returnType);
-    } else if (chosenFunctionName == "row_constructor") {
-      // Avoid generating deeply nested types that is rarely used in practice.
-      if (levelOfNesting(returnType) < 3) {
-        expression = generateRowConstructorExpression(returnType);
-      }
-    } else if (chosenFunctionName == "dereference") {
-      expression = generateDereferenceExpression(returnType);
-    } else {
-      expression = generateExpressionFromConcreteSignatures(
-          returnType, chosenFunctionName);
-      if (!expression &&
-          (options_.enableComplexTypes || options_.enableDecimalType)) {
-        expression = generateExpressionFromSignatureTemplate(
+    auto functionTransformer =
+        options_.functionTransformers.find(chosenFunctionName);
+    if (functionTransformer != options_.functionTransformers.end()) {
+      state.remainingLevelOfNesting_ -=
+          functionTransformer->second->extraLevelOfNesting();
+    }
+
+    if (state.remainingLevelOfNesting_ >= 0) {
+      if (chosenFunctionName == "cast") {
+        expression = generateCastExpression(returnType);
+      } else if (chosenFunctionName == "row_constructor") {
+        // Avoid generating deeply nested types that is rarely used in practice.
+        if (levelOfNesting(returnType) < 3) {
+          expression = generateRowConstructorExpression(returnType);
+        }
+      } else if (chosenFunctionName == "dereference") {
+        expression = generateDereferenceExpression(returnType);
+      } else {
+        expression = generateExpressionFromConcreteSignatures(
             returnType, chosenFunctionName);
+        if (!expression &&
+            (options_.enableComplexTypes || options_.enableDecimalType)) {
+          expression = generateExpressionFromSignatureTemplate(
+              returnType, chosenFunctionName);
+        }
+      }
+    }
+
+    if (functionTransformer != options_.functionTransformers.end()) {
+      if (expression) {
+        expression =
+            functionTransformer->second->transform(std::move(expression));
       }
+      state.remainingLevelOfNesting_ +=
+          functionTransformer->second->extraLevelOfNesting();
```

Comment:
Can we remove the extra logic in L1007 and just add here:
```
  if (expression != nullptr &&
      options_.exprTransformers.count(chosenFunctionName)) {
    auto exprTransformer = options_.exprTransformers.at(chosenFunctionName);
    if (exprTransformer->extraLevelOfNesting() >
        state.remainingLevelOfNesting_) {
      LOG(ERROR)
          << "Expression transformation failed due to too many nesting layers";
      expression = nullptr;
    } else {
      expression = exprTransformer->second->transform(std::move(expression));
    }
  }
```

Reply:
Hi @kewang1024, no, we must do `state.remainingLevelOfNesting_ -= functionTransformer->second->extraLevelOfNesting();` before generating the expression at line 1015--1032. The reason is the following: Suppose the chosen function is `plus`, remainingLevelOfNesting_ at the beginning of the function is 2, and extraLevelOfNesting() is 1. remainingLevelOfNesting_ is first be decremented by 1 at line 974 to exclude the level of the current function call. Then, if we didn't decrement remainingLevelOfNesting_ to be 1-1=0 before line 1015, generateExpressionFromConcreteSignatures() could generate another 1-level subexpression as the argument to `plus`, e.g., `plus(a, multiply(b, c))`. After generateExpressionFromConcreteSignatures() returns back to the current function, remainingLevelOfNesting_ is still 1 and is >= extraLevelOfNesting(), so you would wrap the expression by the wrapper, e.g., `transform_wrapper(plus(a, multiply(b, c)))`. Now, the level of nesting of this expression would exceed the original remainingLevelOfNesting_ = 2.